src/backend/columnar/columnar_reader.c

```diff
@@ -269,39 +276,106 @@ ColumnarReadRowByRowNumber(Relation relation, uint64 rowNumber,
 		return false;
 	}
 
-	TupleDesc relationTupleDesc = RelationGetDescr(relation);
-	List *whereClauseList = NIL;
-	List *whereClauseVars = NIL;
-	MemoryContext stripeReadContext = CreateStripeReadMemoryContext();
-	StripeReadState *stripeReadState = BeginStripeRead(stripeMetadata,
-													   relation,
-													   relationTupleDesc,
-													   neededColumnList,
-													   whereClauseList,
-													   whereClauseVars,
-													   stripeReadContext);
-
-	ReadStripeRowByRowNumber(stripeReadState, stripeMetadata, rowNumber,
-							 columnValues, columnNulls);
+	if (!ColumnarReadIsCurrentStripe(readState, stripeMetadata))
+	{
+		/* do the cleanup before starting to read a new stripe */
+		ColumnarResetRead(readState);
+
+		TupleDesc relationTupleDesc = RelationGetDescr(relation);
+		List *whereClauseList = NIL;
+		List *whereClauseVars = NIL;
+		MemoryContext stripeReadContext = readState->stripeReadContext;
+		readState->stripeReadState = BeginStripeRead(stripeMetadata,
+													 relation,
+													 relationTupleDesc,
+													 readState->projectedColumnList,
+													 whereClauseList,
+													 whereClauseVars,
+													 stripeReadContext);
+
+		/*
+		 * XXX: Since stripes in stripeList are ordered by their first row
+		 * numbers, we can't just assume that stripe with "id" will the "id"th
+		 * item of stripe list.
+		 * For this reason, we should either store StripeMetadata instead of
+		 * currentStripe, or we need to linearly search our stripe to set
+		 * currentStripe, as in below.
+		 */
+		readState->currentStripe = StripeListGetStripeIndex(readState->stripeList,
+															stripeMetadata);
+		Assert(readState->currentStripe >= 0);
```

Comment:
Regarding https://github.com/citusdata/citus/pull/5058#discussion_r655918433, I made some changes in d3e8e18890562cc9e0490f391620abcf20e82a5d. 

On the other hand, above part seems a bit annoying to me.
Now we don't store a redundant `StripeMetadata` object and directly make use of `readState->stripeList[readState->currentStripe]`. 
For this reason, we need to linearly search `stripeList` to correctly set `currentStripe`.

Does that look good ?
Or, should we have another field in `ColumnarReadState` to store `StripeMetadata` itself additionally ?
The second approach would make `ColumnarReadState` slightly confusing, but current approach is not very nice from the performance aspect.

What do you think @jeff-davis ?